### PR TITLE
Replace torch.lstsq with torch.linalg.lstsq

### LINKED ISSUE #3843: Replace  torch.lstsq with torch.linalg.lstsq in `perspective` transform
## 🚀 Feature

We recently started getting the following types of warnings:

```
/root/project/torchvision/transforms/functional.py:597: UserWarning: torch.lstsq is deprecated in favor of torch.linalg.lstsq and will be removed in a future PyTorch release.
torch.linalg.lstsq has reversed arguments and does not return the QR decomposition in the returned tuple (although it returns other information about the problem).
To get the qr decomposition consider using torch.linalg.qr.
The returned solution in torch.lstsq stored the residuals of the solution in the last m - n columns of the returned value whenever m > n. In torch.linalg.lstsq, the residuals in the field 'residuals' of the returned named tuple.
The unpacking of the solution, as in
X, _ = torch.lstsq(B, A).solution[:A.size(1)]
should be replaced with
X = torch.linalg.lstsq(A, B).solution (Triggered internally at  /opt/conda/conda-bld/pytorch_1620975976061/work/aten/src/ATen/LegacyTHFunctionsCPU.cpp:389.)
  res = torch.lstsq(b_matrix, a_matrix)[0]
```

which gets even more visible in [the newly added galleries](https://579199-73328905-gh.circle-artifacts.com/0/docs/auto_examples/plot_transforms.html#sphx-glr-auto-examples-plot-transforms-py)

We should replace
https://github.com/pytorch/vision/blob/b56f17ae1ae8a5d08067c7f7444af21fb3b59ca6/torchvision/transforms/functional.py#L597
with
```python
res = torch.linalg.lstsq(a_matrix, b_matrix).solution
```

So that we stop using a deprecated function.

One caveat is that maybe some tests might start failing (due to its inherent flakiness), and we should fix the tests that break on the way

It would be good to fix this before the next release, so that users don't get those warnings.

cc @vfdev-5

### PR DESCRIPTION
**Pull request by Gaurav7888**
Fixes #3843

Replace torch.lstsq with torch.linalg.lstsq in perspective transform .

It would be good to fix this before the next release, so that users don't get those warnings.

### CODE REVIEW

**Review by datumbox — APPROVED** [review]
@Gaurav7888 welcome to PyTorch community! LGTM.

To merge this you will need to complete the CLA signing (you only need to do it once). Once the tests pass we can merge to master.

File: torchvision/transforms/functional.py
Comment (datumbox) [inline torchvision/transforms/functional.py:596]: You also need to update the following:
```suggestion
    output: List[float] = res.tolist()
```

**Review by datumbox — CHANGES_REQUESTED** [review]
@Gaurav7888 It seems that the two methods do not provide exactly the same results and this means that some of the tests break. See the CI.

To move forwards with this PR we will need to first fix the broken tests. Could you please let me know if you are interested in continuing the work of this PR?

**Comment by Gaurav7888** [comment]
@datumbox  Sure I would like to work upon this PR .

**Comment by datumbox** [comment]
Awesome, we can fix the tests by updating their expected values. It should be an easy fix. :)

**Comment by Gaurav7888** [comment]
@datumbox  I found something useful regarding this issue
Why is the output of torch.lstsq drastically different than np.linalg.lstsq?

torch.lstq(a, b) solves minX L2∥bX−a∥ while np.linalg.lstsq(a, b) solves minX L2∥aX−b∥

a = torch.tensor([[1., 1, 1],
                  [2, 3, 4],
                  [3, 5, 2],
                  [4, 2, 5],
                  [5, 4, 3]])

b = torch.tensor([[-10., -3],
                  [ 12, 14],
                  [ 14, 12],
                  [ 16, 16],
                  [ 18, 16]])

a1 = a.clone().numpy()
b1 = b.clone().numpy()

x, _ = torch.lstsq(a, b)

x1, res, r1, s = np.linalg.lstsq(b1, a1)

print(f'torch_x: {x[:b.shape[1]]}')

print(f'np_x: {x1}')
Results:

torch_x: tensor([[-0.1145, -0.1047, -0.2863],
        [ 0.3591,  0.3372,  0.5407]])
np_x: [[-0.11452514 -0.10474861 -0.28631285]
 [ 0.35913807  0.33719075  0.54070234]]

**Comment by NicolasHug** [comment]
Hi @Gaurav7888 , thanks for your work so far!

The current test failures are due to some instability in `torch.linalg.lstsq`, which actually returns different output across various executions:

```py
import torch
torch.set_printoptions(precision=10)

a = torch.tensor([[    3,     2,     1,     0,     0,     0,     0,     0],
                  [    0,     0,     0,     3,     2,     1,     0,     0],
                  [   32,     3,     1,     0,     0,     0, -1056,   -99],
                  [    0,     0,     0,    32,     3,     1,     0,     0],
                  [   30,    24,     1,     0,     0,     0,  -990,  -792],
                  [    0,     0,     0,    30,    24,     1,  -750,  -600],
                  [    2,    25,     1,     0,     0,     0,     0,     0],
                  [    0,     0,     0,     2,    25,     1,   -50,  -625]]).to(torch.float)
b = torch.tensor([ 0.,  0., 33.,  0., 33., 25.,  0., 25.])
torch.linalg.lstsq(a, b).solution
```

will sometimes output

```
tensor([ 1.0224266052e+00,  4.4453188777e-02, -3.1561851501e+00,
        -3.5440485924e-02,  1.0277750492e+00, -1.9492297173e+00,
        -2.9531884938e-03, -1.8849190092e-03])
```

and sometimes

```
tensor([ 1.0224264860e+00,  4.4453065842e-02, -3.1561782360e+00,
        -3.5440552980e-02,  1.0277750492e+00, -1.9492284060e+00,
        -2.9531936161e-03, -1.8849160988e-03])
```

ultimately leading to a big difference in pixel values.


It's a bit surprising since the matrix is well-conditioned:

```py
torch.linalg.cond(a, p=-2), torch.linalg.cond(a, p=2)
(tensor(0.0003948331), tensor(2532.7155761719))
```

but oh well.

I believe that changing the solver will help here. 'gels' seems to solve the instability and it is the only available driver for CPU and GPU, so let's use it:

```py
torch.linalg.lstsq(a, b, driver='gels').solution
```

I just pushed a fix, hopefully this will make the CI go green!

**Review by NicolasHug — APPROVED** [review]
unittest CI is green so let's merge!

This LGTM and @datumbox had approved this previously, modulo the failing tests which are now resolved.

Thanks @Gaurav7888 !

**Pull request merged**
